### PR TITLE
Generalized Web Encoder APIs

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
The abstractions now operate over char buffers.
The old convenience APIs are now implemented as extensions for these low level abstractions.
Also, I incorporated feedback from ASP.NET team regarding support for DI, performance, and usability.

### CODE REVIEW

File: src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj [hunk 1]
```diff
@@ -15,21 +15,19 @@
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <RunCodeAnalysis>false</RunCodeAnalysis>
+    <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
```
Comment (stephentoub) [inline src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj]: Why these changes?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs [hunk 2]
```diff
@@ -49,46 +50,46 @@ public CodePointFilter(params UnicodeRange[] allowedRanges)
...
+        /// Allows all characters specified by <paramref name="characters"/> through the filter.
         /// </summary>
         /// <returns>
         /// The 'this' instance.
         /// </returns>
-        public CodePointFilter AllowChars(params char[] chars)
+        public CodePointFilter AllowChars(params char[] characters)
         {
-            if (chars != null)
+            if (characters != null)
```
Comment (stephentoub) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: Why is null allowed at all?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs [hunk 3]
```diff
@@ -1,184 +1,134 @@
...
         }
 
-        /// <summary>
-        /// Everybody's favorite HtmlEncode routine.
-        /// </summary>
-        public string HtmlEncode(string value)
+        public override int MaxOutputCharsPerInputChar
         {
-            return _innerUnicodeEncoder.Encode(value);
+            get { return 8; }
```
Comment (stephentoub) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: Could you either move this to a named constant or add a comment explaining where this number comes from?

Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: added comment: // "&#xFFFF;" is the longest encoded form

File: src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj]: Accidentally. I will revert.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: This is an existing behavior. But, maybe we can remove it. @Eilon?

File: src/System.Text.Encodings.Web/tests/IHtmlEncoder.cs [hunk 4]
```diff
@@ -4,12 +4,12 @@
 using System;
 using System.IO;
 
-namespace System.Text.Encodings.Web
+namespace Microsoft.Framework.WebEncoders
```
Comment (akoeplinger) [inline src/System.Text.Encodings.Web/tests/IHtmlEncoder.cs]: Why the namespace change?

File: src/System.Text.Encodings.Web/tests/PerformanceTests.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text.Encodings.Web;
+using Xunit;
+using Xunit.Abstractions;
+
+#if RELEASE
```
Comment (akoeplinger) [inline src/System.Text.Encodings.Web/tests/PerformanceTests.cs]: I suppose these tests are taking long, we have the `[Outerloop]` attribute for those which I think is nicer than the ifdef.

File: src/System.Text.Encodings.Web/tests/IHtmlEncoder.cs (hunk 4 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/tests/IHtmlEncoder.cs]: Because this interface won't be included in CoreFx.

File: src/System.Text.Encodings.Web/tests/PerformanceTests.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/tests/PerformanceTests.cs]: I did not do it because they take long. I did that so people don't compare debug perf results :-)

**Comment by akoeplinger** [comment]
I mentioned on the initial PR that some of these APIs are general-purpose and should be moved out of System.Text.Encodings.Web, looks some of it was adressed by moving types to the System.Text.Unicode namspace. Is the plan here to move them to a separate package as well? Why was CodePointFilter not moved?

File: src/System.Text.Encodings.Web/tests/PerformanceTests.cs (hunk 5 above)
Comment (akoeplinger) [inline src/System.Text.Encodings.Web/tests/PerformanceTests.cs]: Clever :) I think I'd prefer a special ifdef `BENCHMARK` that is set in the csproj for Release configs though.

**Comment by KrzysztofCwalina** [comment]
@akoeplinger, as I answered before, @ellismg will be doing this refactoring.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharsBitmap.cs [hunk 6]
```diff
@@ -21,9 +21,9 @@ private AllowedCharsBitmap(uint[] allowedCharsBitmap)
         // Marks a character as allowed (can be returned unencoded)
         public void AllowCharacter(char c)
         {
-            uint codePoint = (uint)c;
-            int index = (int)(codePoint >> 5);
-            int offset = (int)(codePoint & 0x1FU);
+            int codePoint = c;
+            int index = codePoint >> 5;
```
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharsBitmap.cs]: Is this computation safe with two's complement? My two's complement knowledge is super rusty (I learned that what, like, 15 years ago?? :smile: ). That is, for "very large" numbers that are positive in `uint`, but negative as `int`, will `>>` maintain the sign after the shift? E.g. does `1000 0000 0000 0000 >> 5` become `1111 1100 0000 0000` or `0000 0100 0000 0000`?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: I suppose it's if you, for example, are chaining this in a big call, and querying other components to get their lists of char's, some of those other components might return null and there's no harm in that.

Or to turn it around, why *disallow* null?

Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: Though it's worth pointing out that some of the other methods here are not null-safe, such as the immediately following `AlloChars(string)` that accepts a null string, but will immediately NullRef.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HexUtil.cs [hunk 7]
```diff
@@ -23,6 +23,16 @@ internal static char IntToChar(uint i)
         }
 
         /// <summary>
+        /// Converts a number 0 - 15 to its associated hex character '0' - 'F'.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToHexDigit(int value)
```
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HexUtil.cs]: This should sooooooo be a BCL method! As in a public BCL method on `System.Char`, either static or instance.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs [hunk 8]
```diff
@@ -1,184 +1,128 @@
...
-        /// </summary>
-        public HtmlEncoder(ICodePointFilter filter)
-            : this(new HtmlUnicodeEncoder(CodePointFilter.Wrap(filter)))
+        public static HtmlEncoder Create(params UnicodeRange[] allowedRanges)
         {
+            return new DefaultHtmlEncoder(allowedRanges);
         }
+    }
+
+    public sealed class DefaultHtmlEncoder : HtmlEncoder
```
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: For aspnet/EF we always do one top-level type per file to make it easier to navigate the source tree. In VS it doesn't matter so much, but navigating in GitHub is easy if the filename matches the type. While in GitHub you can press <kbd>t</kbd> and type part of the filename to quickly jump to a file in the repo.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs [hunk 9]
```diff
@@ -1,184 +1,128 @@
...
-                {
-                    writer.Write(chars[--numCharsWritten]);
-                } while (numCharsWritten != 0);
-                writer.Write(';');
+                *buffer = chars[--numCharsWritten];
+                buffer++;
             }
+            while (numCharsWritten != 0);
+
+            *buffer = ';';
```
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: What a weird line of code! :smile:

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+                            buffer++;
+                            bufferSize--;
+                            totalWritten++;
+                        }
+                        else
+                        {
+                            int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, secondChar, out wasSurrogatePair);
+                            if (!encoder.TryEncodeUnicodeScalar(nextScalar, buffer, bufferSize, out charsWritten))
+                            {
+                                throw new ArgumentException("encoder"); // the encoder MaxOutputCharsPerInputChar is not implemented correctly
```
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs]: FxCop #fail.

**Comment by Eilon** [comment]
On general comments:

The original code did some fancy-looking stuff with `Interlocked.CompareExchange` and some `volatile` reads and whatnot. This PR seems to change some of that - are there any negative consequences to that? Or did the code really just move around a bit?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs (hunk 8 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: Yeah, I totally agree for framework sources. Note that these are just for tests. Moreover, once we move APS to the new APIs, these types will be removed in favor of regression performance tests (currently they serve as perf baselines).

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs (hunk 9 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: It does look funny.

**Comment by KrzysztofCwalina** [comment]
I removed the interlocked/volatile operations. The consequences are the following:
1. Once in a blue moon more than one instance of what amounts to a stateless singleton will be created. The last one will win and will be assigned to the static field (for good), the other will be GCed.
2. The new code is faster when accessing the default encoders as it does not have to synchronize caches. 

To me this seems like the right tradeoff.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HexUtil.cs (hunk 7 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HexUtil.cs]: @ellismg will be doing Unicode/char cleanup here. I filed issue #1647 for him to look at these APIs.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs (hunk 10 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs]: Could you clarify?

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: The reason to disallow null here is that it probably is a bug in calling code if null is passed.
Having said that, I don't feel super strongly about it, so I am ok leaving as is.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharsBitmap.cs (hunk 6 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharsBitmap.cs]: The argument starts as a char (16 bit unsigned). It's Max value will be 65K. When it's cast to an int the value is never negative (and fits in 32 bits just fine), so right shifts work as expected.

Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/AllowedCharsBitmap.cs]: Ah, good point.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: Ultimately I think the bigger issue is that the public entry points to the various encoding-related APIs are not fully consistent:

1. Some do null checks and no-op (this method)
2. Some do null checks and throw (various other APIs, such as `TextEncoderExtensions.Encode(...)`)
3. Some don't do null checks and NullRef (such as `AllowChars(string)` right below this method).

I agree that between 1. and 2. it's debatable and really depends on the scenario, but even within the scope of this change I think it might be inconsistent even for particular scenarios. E.g. what's so wrong about a `null` value for a `value` parameter? Sometimes you're getting the value from "some other place" that might return null when there's no data, but you don't care, you just want it to be safe to render out as HTML/JS/URL so you just want to get back `""` or whatever. For some methods it totally makes sense to throw: e.g. you called an API that requires passing in a particular encoding, and you passed in `null` - the operation can't possibly succeed, and it's clearly a bug in the app's code.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs (hunk 8 above)
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/HtmlEncoder.cs]: Got it, that's fine.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs (hunk 10 above)
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs]: The `ArgumentException` ctor that takes one param takes it as a message, not as a parameter name: https://msdn.microsoft.com/en-us/library/wtxc6334(v=vs.110).aspx

**Comment by Eilon** [comment]
Re: interlocked/volatile - Thanks, that makes sense to me but I have to admit it's not my area of expertise. Just pinging @grabyourpitchforks here in case he has any last thoughts on that.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs (hunk 10 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs]: Ah, thanks. good catch. I will fix that.

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: I agree that it would be good to look at these APIs and clean that up. But I propose I do it as a separate workitem. The goal of this PR is to generalize the abstraction APIs, and implement the same behavior like in current ASP.NET implementation. Cleaning up null behavior seems like a totally separate and potentially large workitem.

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
@dotnet-bot test this please

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (Eilon) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: No problem with the cleanup happening in a separate PR, but I feel it does have to happen before shipping. Changing it later (depending on the change) would be a breaking change. And... we all know these things often get "forgotten" or punted if they're left too late. Totally your call.

**Comment by Eilon** [comment]
:shipit:

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/CodePointFilter.cs]: This one won't be forgotten. #1678.

**Comment by benaadams** [comment]
How would you encode HTML/Javascript/json directly to byte[] or byte* in UTF8 or similar for a network or disk? Or would you have to encode to char/string; then re-encode to byte in double pass?
